### ISSUE TITLE
Crash bug with specific artist

### ISSUE THREAD

**Issue by L0wnWolf** [apsun/Cardapter#2]
![Screenshot_20200909-192142](https://user-images.githubusercontent.com/41022693/92665790-71d59500-f2d5-11ea-82b0-2686ef078391.jpg)
![Screenshot_20200909-191308](https://user-images.githubusercontent.com/41022693/92665792-726e2b80-f2d5-11ea-8861-4056228105ea.jpg)

Cardapter is crashing with the specific artist in the screenshots(at least those 2 specific songs for sure). Once Cardapter tries to play either of those 2 songs, Android Auto crashes entirely. It tries to reconnect several times and then leaves the phone disconnected from the car. If I then skip the song and reconnect the phone, it will work as normal. My only hypothesis would be if Cardapter is set to another country than what I'm in and it creates a conflict where the song is not available to the Cardapter country and cannot play. I'd be happy to help further troubleshoot this

**Comment by apsun** [comment]
:o That's a pretty interesting bug! Can you grab a logcat dump containing the crash? (using `adb logcat` on a computer, or an app like MatLog on your phone)

If I had to guess I would say it's the emoji in the artist name, I can't imagine what else would cause this. Do all of their songs crash Android Auto?

**Comment by L0wnWolf** [comment]
I went out and tried 3 more songs from the same artist. Cardapter crashed every time. When I played a song from a different artist there was no crash. I have attached the log output for the crash.
![crash_log](https://user-images.githubusercontent.com/41022693/92769303-d42aa600-f366-11ea-970d-c280bd9d38d0.PNG)

**Comment by L0wnWolf** [comment]
I found another song from a different artist that is crashing where this one has some bold text in the song name. I think you're right where the issue lies with when there is either an emoji or text that is not standard.
![Screenshot_20200910-141802](https://user-images.githubusercontent.com/41022693/92780399-178a1200-f371-11ea-9330-a5e830c649d0.jpg)

**Comment by apsun** [comment]
Hmm weird I can't reproduce this on my phone. Do you have any error logs before those lines? The lines you posted seem to be a symptom of the crash, not a cause.

**Comment by L0wnWolf** [comment]
![crash_log2](https://user-images.githubusercontent.com/41022693/93026270-e01aa000-f5d2-11ea-8937-98152f765027.PNG)
[logcat.txt](https://github.com/apsun/Cardapter/files/5214956/logcat.txt)

I went and caused it to crash again and pulled more logs. The attached photo seems to be where it crashed. I'm also pasting in a few extra lines below that I'm not sure if are useful or not but figured I'd mention them here. I also added the full log file in case there's something I missed. Let me know if I can do anything else

09-13 14:14:18.042 22154 22154 E ActivityThread: Service com.crossbowffs.cardapter.CardapterBrowserService has leaked ServiceConnection android.media.browse.MediaBrowser$MediaServiceConnection@3f4158e that was originally bound here

09-13 14:47:16.845  2207  2207 D OPSystemUIWLBHelper: packageName com.soundcloud.android mCurrentMode 0 isMediaAllowed false

09-13 14:47:23.813 11723 12499 I AddressedMediaPlayer_ext: null queue from com.soundcloud.android, constructing single-item list

**Comment by apsun** [comment]
Thanks for the logs! Unfortunately I don't think they show why the app is actually crashing, or really who/what is responsible for the crash. I made a build with extra logging and a failsafe to try to prevent Android Auto from falling over in case it's SoundCloud that's causing the issue, mind giving this a try?

[app-release.zip](https://github.com/apsun/Cardapter/files/5215673/app-release.zip)

**Comment by L0wnWolf** [comment]
I tried the new apk and it still results in a crash. I'm adding the updated crash log, but I wasn't able to find anything that stood out as the actual crash this time. Is there another log I can pull besides the crash log?
[logcat.txt](https://github.com/apsun/Cardapter/files/5215907/logcat.txt)

**Comment by apsun** [comment]
I don't see any logs from the Cardapter app, how are you obtaining these logs? The last reference I see to Cardapter in the logs are some install events.

Does the crash happen with the phone version of Android Auto (i.e. if you open the app without connecting to your car)? If you are able to reproduce the crash with the phone version, I'd recommend having your phone connected to the PC (and adb logcat running) while you cause the crash. Otherwise, maybe use MatLog or some other app that is able to read the logs on the device in real time.

**Comment by L0wnWolf** [comment]
![Screenshot_20200913-234210](https://user-images.githubusercontent.com/41022693/93041672-0f54ff80-f61b-11ea-850c-ec7b9579bea6.jpg)

It does not happen when using AA on the phone itself. Before I was just using a command on pc to obtain crash logs after the fact. I went and got a log app and caused it to crash again. The 2 screenshots are the 2 immediate logs I got as the crash happened

**Comment by L0wnWolf** [comment]
![Screenshot_20200913-234206](https://user-images.githubusercontent.com/41022693/93041837-87232a00-f61b-11ea-92e4-5884652565f3.jpg)
First photo didn't post

**Comment by apsun** [comment]
Interesting. From the error you posted, this seems to be a... bug? unintended feature? of the SoundCloud app. I think what happened here is since SoundCloud doesn't intend to be called from outside the app, they put internal objects that only can be deserialized by the SoundCloud app into the song metadata. When Android Auto tries to read this, it doesn't know how to handle it and crashes.

So then the questions are:

1. Why does it only happen on certain artists?
2. Why does it only happen in the car UI?

No idea how to answer these questions, unfortunately.

As for a workaround on our end, I think it is possible, but we would basically need to intercept all the communication between Android Auto and SoundCloud. This is actually surprisingly hard (it was how I originally tried to make the app but eventually I gave up and went with the hacky approach we currently use).

tl;dr: SoundCloud app sends bad data, Android Auto doesn't handle bad data, we can only sit helplessly and watch :-(

**Comment by L0wnWolf** [comment]
One possible thought is since it seems to be triggered by artists or song names with special characters/emoji, is there a way to add something where Cardapter only accepts standard text and either deletes or ignores the rest?

**Comment by apsun** [comment]
That's the problem, we don't sit in between AA SoundCloud. It's more like, we take the AA socket and the SC plug and we connect them together. Since we don't sit in the middle we can't change what AA receives.

This is why I said we would need to significantly change the way Cardapter works if we want to do this. :-(
